docs/adrs/0001-support-caching-deps-for-monorepos.md

```diff
@@ -0,0 +1,48 @@
+# 0. Support caching dependencies for mono repos
+Date: 2021-07-13
+
+Status: Proposed
+
+## Context
+Currently, `actions/setup-node` supports caching dependencies for Npm and Yarn package managers.  
+For the first iteration, we have decided to not support cases where `package-lock.json` / `yarn.lock` are located outside of repository root.  
+Current implementation searches the following file patterns in the repository root: `package-lock.json`, `yarn.lock` (in order of resolving priorities)
+
+Obviously, it made build-in caching unusable for mono-repos and repos with complex structure.  
+We would like to revisit this decision and add customization for dependencies lock file location.
+
+## Proposal
+We have the following options:
+1. Allow to specify  directory where `package-lock.json` or `yarn.lock` are located
+2. Allow to specify path to the dependencies lock file (including directory path and filename)
+
+The second option looks more generic because it allows to:
+- specify multiple dependencies files using file patterns like `**/package-lock.json` ([one of recommended approaches in actions/cache](https://github.com/actions/cache/blob/main/examples.md#macos-and-ubuntu))
+- specify custom dependencies files like `src/npm-shrinkwrap.json`
+- change default resolving priority if both `yarn.lock` and `package-lock.json` exist in repository
+
+## Decision
+
+Add `cache-dependency-path` input that will accept path (relative to repository root) to dependencies lock file.  
+If provided path contains wildcards, the action will search all maching files and calculate common hash like `${{ hashFiles('**/package-lock.json') }}` YAML construction does.  
```

Comment:
How is common hash helping the situation? I think it makes it worse, because if I have monorepo with 10 projects and only one `package-lock.json` file is changed, entire cache is not usable. Or am I missing something?

Reply:
The common hash is only one of possible options. You can also specify path to the single file.
Let's imagine that you have the following repo structure:
```
- service1
--- package.json
--- package-lock.json
--- ...
- service2
--- package-lock.json
--- ...
```

In your first workflow, where you build `service1`, you will invoke setup-node as
```yml
- uses: actions/setup-node@v2
  with:
    cache: npm
    cache-dependency-path: service1/package-lock.json
```
In the second workflow, you will invoke setup-node as
```yml
- uses: actions/setup-node@v2
  with:
    cache: npm
    cache-dependency-path: service2/package-lock.json
```

So every workflow will use own cache and if one `package-lock.json` is changed, only one cache will be regenerated. 

Option with wildcard makes sense when you build multiple services in scope of one workflow, so you will have common cache for a few services. This approach is recommended here: https://github.com/actions/cache/blob/main/examples.md#macos-and-ubuntu

ADR contains a bunch of other examples.
Hope I answered your question.

Reply:
Yes, thank you very much. I am looking forward to it, looks pretty amazing :)

Reply:
Just curious... why name it `cache-dependency-path` and not something like `cache-lock-file-glob`?

It isn't really a path to dependencies is it?

Reply:
https://github.com/actions/setup-node/pull/299#discussion_r669502996

Reply:
Okay, interesting.

Why make it generic to accommodate python? This is `setup-node`... Is there a plan to make this action handle caching across all languages? If not I would argue that being generic here leads to confusion. It's not obvious from the name that you're looking for the location of my lock files (as opposed to `package.json` files, or `node_modules` folders).

Also, path doesn't really handle what is being done here... you're using a glob. This can cover multiple files _and multiple paths_.

Reply:
Thank you for sharing your opinion. We prefer to use `cache-dependency-path` as the input name for the reasons listed in [this comment](https://github.com/actions/setup-node/pull/299#discussion_r669502996).